### PR TITLE
[SPARK-37478][SQL][TESTS][FOLLOWUP] Unify v1 and v2 DROP NAMESPACE error

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
According to [#cmt](https://github.com/apache/spark/pull/34819#discussion_r763629569), unify the error of v1, v2 and hive external catalog in DROP NAMESPACE tests.

### Why are the changes needed?
Currently, v1 and hive external catalog command throw `AnalysisException`, while v2 command throws `SparkException`. The error messages of v1 and hive catalog are also completely different from v2.
So this PR is for unifying those errors to `AnalysisException`. The error message will have one difference between v1/hive and v2: `Cannot drop a non-empty database: ...` vs `Cannot drop a non-empty namespace: ...`

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
v1/v2 and Hive v1 DropNamespaceSuite:

```$ build/sbt -Phive-2.3 -Phive-thriftserver "test:testOnly *DropNamespaceSuite"```

### CODE REVIEW

**Comment by dchvn** [comment]
cc @cloud-fan @huaxingao. FYI !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -540,8 +540,14 @@ object QueryCompilationErrors {
       s"rename temporary view from '$oldName' to '$newName': destination view already exists")
   }
 
-  def databaseNotEmptyError(db: String, details: String): Throwable = {
-    new AnalysisException(s"Database $db is not empty. One or more $details exist.")
+  def cannotDropNonemptyDatabaseError(db: String, details: String): Throwable = {
+    new AnalysisException(s"Cannot drop a non-empty database: $db. One or more $details exist.")
+  }
```
Comment (dchvn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:545]: Change the error message of v1 command.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -540,8 +540,14 @@ object QueryCompilationErrors {
...
+
+  def cannotDropNonemptyNamespaceError(namespace: Seq[String]): Throwable = {
+    new AnalysisException(
+      s"Cannot drop a non-empty namespace: ${namespace.quoted}. " +
+        "Use CASCADE option to drop a non-empty namespace.")
```
Comment (dchvn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:550]: Move `cannotDropNonemptyNamespaceError` from `QueryExecutionErrors` to `QueryCompilationErrors` as [SPARK-33539](https://issues.apache.org/jira/browse/SPARK-33539)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DropNamespaceSuiteBase.scala [hunk 3]
```diff
@@ -36,7 +36,7 @@ trait DropNamespaceSuiteBase extends QueryTest with DDLCommandTestUtils {
 
   protected def builtinTopNamespaces: Seq[String] = Seq.empty
   protected def isCasePreserving: Boolean = true
-  protected def assertDropFails(): Unit
+  protected def namespaceAlias: String = "namespace"
```
Comment (dchvn) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DropNamespaceSuiteBase.scala:39]: Alias for namespace: `database` for v1 and `namespace` for v2

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 4]
```diff
@@ -196,7 +196,15 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
-    client.dropDatabase(db, ignoreIfNotExists, cascade)
+    try {
+      client.dropDatabase(db, ignoreIfNotExists, cascade)
+    } catch {
+      case NonFatal(exception) =>
+        if (exception.getClass.getName.equals("org.apache.hadoop.hive.ql.metadata.HiveException")
+          && exception.getMessage.contains(s"Database $db is not empty.")) {
+          throw new AnalysisException(s"Cannot drop a non-empty database: $db")
+        } else throw exception
+    }
```
Comment (dchvn) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:207]: Try-catch the error from hive catalog and re-throw it with a different error message that matches v1, v2

**Comment by dchvn** [comment]
Please take a look if you have time @cloud-fan, @huaxingao . Many thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:550]: Does CASCADE work for the v1 command? If yes, we should also mention it in the v1 error message.

Comment (dchvn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:550]: Yes, it works. I'll mention it in v1 error message. Thank you, @cloud-fan!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 5]
```diff
@@ -540,8 +540,14 @@ object QueryCompilationErrors {
       s"rename temporary view from '$oldName' to '$newName': destination view already exists")
   }
 
-  def databaseNotEmptyError(db: String, details: String): Throwable = {
-    new AnalysisException(s"Database $db is not empty. One or more $details exist.")
+  def cannotDropNonemptyDatabaseError(db: String, details: String): Throwable = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:543]: Since we can remove the `details` parameter now.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 5 above)
Comment (dchvn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:543]: Sorry, I missed the call `cannotDropNonemptyDatabaseError(db, "tables")` and `cannotDropNonemptyDatabaseError(db, "functions")` in `InMemoryCatalog` file. So I think keep origin error message and also mention `use CASCADE`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:543]: Actually, I think `One or more $details exist` is a repeat of `non-empty database`, and it's OK to remove it, to keep consistent with v2.

Comment (dchvn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: Updated. Thank you @cloud-fan!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 6]
```diff
@@ -196,7 +196,15 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
       ignoreIfNotExists: Boolean,
       cascade: Boolean): Unit = withClient {
-    client.dropDatabase(db, ignoreIfNotExists, cascade)
+    try {
+      client.dropDatabase(db, ignoreIfNotExists, cascade)
+    } catch {
+      case NonFatal(exception) =>
+        if (exception.getClass.getName.equals("org.apache.hadoop.hive.ql.metadata.HiveException")
+          && exception.getMessage.contains(s"Database $db is not empty.")) {
+          throw new AnalysisException(s"Cannot drop a non-empty database: $db.")
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:205]: shall we use `cannotDropNonemptyDatabaseError`?

Comment (dchvn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Updated. Thank you @cloud-fan!

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**